File: py800mon/monitor/statusupdater.py
import asyncio
import time

from ..actions import Actions
from ..app import EventType
from ..datastructures import CpuState
from ..atari.disasm import disasm_6502_one
from ..emulator import CAP_MONITOR_BREAKPOINTS
from ..rpc import RpcException
from .appstate import state


class StatusUpdater:
    def __init__(
        self,
        rpc,
        dispatcher,
        paused_interval=1.0,
        running_interval=0.05,
        error_interval=1.0,
    ):
        self._rpc = rpc
        self._dispatcher = dispatcher
        self._paused_interval = float(paused_interval)
        self._running_interval = float(running_interval)
        self._error_interval = float(error_interval)
        self._wake_event = asyncio.Event()
        self._caps_synced = False
        self._last_caps_attempt = 0.0

    def request_refresh(self):
        self._wake_event.set()

    async def run(self, event_queue):
        force_cpu_refresh = False
        while True:
            await self._poll_once(force_cpu_refresh)
            force_cpu_refresh = False
            await event_queue.put((EventType.STATUS, None))
            interval = (
                self._error_interval
                if state.last_rpc_error
                else self._paused_interval if state.paused else self._running_interval
            )
            try:
                await asyncio.wait_for(self._wake_event.wait(), timeout=interval)
                self._wake_event.clear()
                force_cpu_refresh = True
            except asyncio.TimeoutError:
                pass

    async def _poll_once(self, force_cpu_refresh=False):
        had_error = state.last_rpc_error != ""
        try:
            status = await self._rpc.status()
        except RpcException:
            self._sync_rpc_error()
            return

        changed = (
            state.paused != status.paused
            or state.emu_ms != status.emu_ms
            or state.reset_ms != status.reset_ms
            or state.crashed != status.crashed
            or state.state_seq != status.state_seq
        )
        if changed:
            self._dispatcher.dispatch(Actions.SET_STATUS, status)
        if changed or force_cpu_refresh:
            await self._update_cpu()
        now = time.monotonic()
        need_caps = had_error or not self._caps_synced
        if need_caps and now - self._last_caps_attempt >= 1.0:
            self._last_caps_attempt = now
            await self._update_capabilities()
        self._sync_rpc_error()

    async def _update_cpu(self):
        try:
            data = await self._rpc.cpu_state()
        except RpcException:
            return

        ypos, xpos, pc, a, x, y, s, p = data
        cpu = CpuState(ypos=ypos, xpos=xpos, pc=pc, a=a, x=x, y=y, s=s, p=p)
        cpu_disasm = ""
        try:
            code = await self._rpc.read_memory(pc, 3)
            cpu_disasm = disasm_6502_one(pc, code)
        except (RpcException, RuntimeError):
            pass
        self._dispatcher.dispatch(Actions.SET_CPU, (cpu, cpu_disasm))

    def _sync_rpc_error(self):
        error = self._rpc.last_error
        text = str(error) if error else None
        if state.last_rpc_error != text:
            self._dispatcher.dispatch(Actions.SET_LAST_RPC_ERROR, text)

    async def _update_capabilities(self):
        try:
            caps = await self._rpc.config()
        except RpcException:
            return
        self._caps_synced = True
        supported = CAP_MONITOR_BREAKPOINTS in set(caps)
        if state.breakpoints_supported != supported:
            self._dispatcher.dispatch(
                Actions.SET_BREAKPOINTS_SUPPORTED,
                supported,
            )

File: go800mon/a800mon/monitor/statusupdater.go
package monitor

import (
	"context"
	"time"

	. "go800mon/a800mon"
	"go800mon/internal/disasm"
)

type StatusUpdater struct {
	rpc             *RpcClient
	dispatcher      *ActionDispatcher
	pausedInterval  time.Duration
	runningInterval time.Duration
	errorInterval   time.Duration
	lastPoll        time.Time
	forceRefresh    bool
	capsSynced      bool
	lastCapsAttempt time.Time
}

func NewStatusUpdater(rpc *RpcClient, dispatcher *ActionDispatcher, pausedInterval, runningInterval time.Duration) *StatusUpdater {
	return &StatusUpdater{
		rpc:             rpc,
		dispatcher:      dispatcher,
		pausedInterval:  pausedInterval,
		runningInterval: runningInterval,
		errorInterval:   time.Second,
	}
}

func (s *StatusUpdater) RequestRefresh() {
	s.forceRefresh = true
}

func (s *StatusUpdater) Tick(ctx context.Context) (bool, error) {
	st := State()
	hadError := st.LastRPCError != ""
	interval := s.runningInterval
	if st.Paused {
		interval = s.pausedInterval
	}
	if st.LastRPCError != "" {
		interval = s.errorInterval
	}
	if !s.forceRefresh && !s.lastPoll.IsZero() && time.Since(s.lastPoll) < interval {
		return false, nil
	}
	forced := s.forceRefresh
	s.lastPoll = time.Now()
	s.forceRefresh = false

	status, err := s.rpc.Status(ctx)
	if err != nil {
		s.syncRPCError()
		return true, nil
	}
	changed := st.Paused != status.Paused ||
		st.EmuMS != status.EmuMS ||
		st.ResetMS != status.ResetMS ||
		st.Crashed != status.Crashed ||
		st.StateSeq != status.StateSeq
	if changed {
		_ = s.dispatcher.Dispatch(ActionSetStatus, status)
	}
	if changed || forced {
		s.updateCPU(ctx)
	}
	needCaps := hadError || !s.capsSynced
	if needCaps {
		now := time.Now()
		if hadError || s.lastCapsAttempt.IsZero() || now.Sub(s.lastCapsAttempt) >= time.Second {
			s.lastCapsAttempt = now
			if s.updateCapabilities(ctx) {
				s.capsSynced = true
			}
		}
	}
	s.syncRPCError()
	return true, nil
}

func (s *StatusUpdater) updateCPU(ctx context.Context) {
	cpu, err := s.rpc.CPUState(ctx)
	if err != nil {
		return
	}
	cpuDisasm := ""
	if code, err := s.rpc.ReadMemory(ctx, cpu.PC, 3); err == nil {
		cpuDisasm = disasm.DisasmOne(cpu.PC, code)
	}
	_ = s.dispatcher.Dispatch(
		ActionSetCPU,
		CPUUpdate{CPU: cpu, Disasm: cpuDisasm},
	)
}

func (s *StatusUpdater) syncRPCError() {
	err := s.rpc.LastError()
	text := ""
	if err != nil {
		text = err.Error()
	}
	if State().LastRPCError != text {
		_ = s.dispatcher.Dispatch(ActionSetLastRPCError, text)
	}
}

func (s *StatusUpdater) updateCapabilities(ctx context.Context) bool {
	caps, err := s.rpc.Config(ctx)
	if err != nil {
		return false
	}
	supported := false
	for _, id := range caps {
		if id == CapMonitorBreakpoints {
			supported = true
			break
		}
	}
	if State().BreakpointsSupported != supported {
		_ = s.dispatcher.Dispatch(ActionSetBreakpointsSupported, supported)
	}
	return true
}
